Refactoring the LoggingManager - V1.0

Vincent-SD opened the pull request:
See med-material/LoggingManager#13

* Adding the logStore class
   - Ability to create the log string on the fly or at the end of the logging process by setting a boolean to true or false in Unity.
* Adaptation of the LoggingManager
* Adding the WriteToCSV class

The MySQL part is not yet implemented and adapted in this version.

bastianilso commented:
OK, thanks for showing me the state of things Vincent 🥥.

I think we should structure the pull requests the following way:
- you continue working on the implementation in ArduinoLogger, if this is what is most comfortable for you (and a nice testing bed).
- We then test that the LoggingManager works in ArduinoLogger, Whack-A-Mole VR and HandStrengthener.
- After this is confirmed, I'd like you to submit a pull request to the LoggingManager repo itself.
- Ideally, we structure your changes the following way: Commit 1) Change Dictionary to list structure, Commit 2) Create LogStore class, 3) Convert CSV Writing to WriteToCSV class, 4) Convert SQL class, 5) style fixes and misc. Note, the point is not necessarily that you make the code work between each commit, but instead that it becomes easier for me to review your code.
- After upstream LoggingManager is updated, we make a new release and Unity package.
- The applications using LoggingManager can then update by importing this Unity package and override the existing implementation.

Vincent-SD commented:
I agree, I am currently working on the loggingManager to make it work in Whack-A-Mole VR with Quentin. 
After that i will see if it is working too with HandStrengthener and I will restructure the commits.

Vincent-SD commented:
I decided to separate this pull request in 3 commits.

## Structure of the LoggingManager 

In this commit, I created the structure of the new LoggingManager by creating the new classes :
### Logstore
This class will be used to store the logs and organize them in the correct format. 
Then, it will be possible to export the logs in a string format or in a dictionary format (see [med-material/LoggingManager#13](https://github.com/med-material/LoggingManager/issues/13)).

### WriteToCSV

This class will allow to save the logs in a file. 

## Implementation of LogStore class

This is the implementation of all the functions in the LogStore class. Here are useful things to know about this class.
* The first row logged will define the headers of the logs
* It is possible to enable/disable the function that creates the string on the fly (during the log process)
	* If an header is added during the log process  :
		* A new header is created and the "NULL" value is put for all the previous rows
		* If the log string on the fly functionality is enabled, it will be disabled
 ![logs](https://user-images.githubusercontent.com/49280157/145183448-374bd521-7d05-4458-9bcf-10a02309afbd.png)

* Each time a line is fully logged, the TerminateRow() function has to be called. In order for this new version to work with previous implementations of the LoggingManager, this function is called automatically when needed (as explained [here](https://github.com/Vincent-SD/ArduinoLogger/blob/85cc54eba8d6a4569b986281ca28df14de1d717b/Assets/LoggingManager/LogStore.cs#L91)).

## Implementation of LogToCSV class

Each LogToCSV object is instantiated with a file path and a LogStore object. 
When calling the WriteAll() function, the logs are exported from the LogStore object as a string and are written in the file at the given file path. The process is currently executed in a separated thread to prevent the main application from freezing.

Vincent-SD commented:
@bastianilso I can't request review for some unknown reason, so I'm commenting on this pull request🌴

bastianilso commented:
@Vincent-SD great! One last request: could you split the your changes to LoggingManager.cs and LoggingExample.cs into a separate commit that happens before the changes to ArduinoLogger? ❄

bastianilso reviewed (changes requested):
🌏

bastianilso commented on Assets/LoggingManager/LogStore.cs:
```diff
@@ -0,0 +1,276 @@
+    private int nbLines;
+    private bool createStringOverTime;
+    private StringBuilder currentLineLogged;
+    private SortedDictionary<string, string> currentLogs;
```
Whats the difference between `logs` and `currentLogs`  dictionaries ? - not sure I understand the difference just from the naming convention alone. And it seems like logs are added to both dictionaries in e.g. `Add()` ?

EDIT: I understand now. I think you could maybe call it `currentRow` or `currentLogRow` to make it clear its relationship also to your function `TerminateRow`. (which we could also call `EndRow` i guess)

Vincent-SD replied:
The `logs` dictionary contains all the logs of each row while the `currentsLogs` dictionary contains only the logs of the current line.
I think we don't really need to rename `logs` because this is just the dictionary where all the logs are stored, what do you think ?
I agree with your second proposal, let's rename `currentLogs` to `currentLogRow` and `TerminateRow` to `EndRow`, it will make more sense.

bastianilso replied:
I think `logs` is a fine name. Thanks for the explanation.

bastianilso commented on Assets/LoggingManager/LogStore.cs:
```diff
@@ -0,0 +1,276 @@
+    private StringBuilder currentLineLogged;
+    private SortedDictionary<string, string> currentLogs;
+    private const string fieldSeparator = ";";
+    private const string lineSeperator = "\n";
```
should be `lineSeparator`. (replace e with a)

Vincent-SD replied:
Oops 😅

bastianilso commented on Assets/LoggingManager/LogStore.cs:
```diff
@@ -0,0 +1,276 @@
+    private string email;
+    public string SessionId { get; set; }
+
+    public LogStore(string label, string email, bool createStringOverTime,
+        LogType logType = LogType.Normal, bool logCommonColumns = true)
+    {
+        Init(label, email, Guid.NewGuid().ToString(), createStringOverTime, logType, logCommonColumns);
+    }
```
Could we set the `createStringOverTime` bool to `true` by default? Would that work for e.g. Whack-A-Mole etc.? Either way it would be good to settle on a default value, so people dont have to think of it.

What is the `Guid.NewGuid()` function? Is it some built-in way to generate IDs? What is it using internally?

Vincent-SD replied:
### "Could we set the createStringOverTime bool to true by default?"
The `createStringOverTime` bool is set to true in the LoggingManager [here](https://github.com/Vincent-SD/ArduinoLogger/blob/06cbd9a1d1288205624b9d286d0c12b86799b5c6/Assets/LoggingManager/LoggingManager.cs#L35).

The user don't have to pay attention to this value because it is already defined in the manager and the LogStore objects are only created by this manager. I decided not to put a default value in the LogStore class to make the code more adaptable if one day it is decided to put some parts of this code in another project.

### "Would that work for e.g. Whack-A-Mole etc.?"
It will not work everywhere.  The log structure must be clean and the program cannot add headers during the logging process, otherwise this feature will be disabled. 

### "What is the Guid.NewGuid() function? Is it some built-in way to generate IDs? What is it using internally?"

This function is used to generate a random unique ID, so that each log will have its own id even if the user doesn't give it one.

bastianilso replied:
Fine with me, thanks for the explanation.
Also, fine with using `Guid.NewGuid()`, I think it's better than us making up our own way of making IDs.

bastianilso commented on Assets/LoggingManager/LogStore.cs:
```diff
@@ -0,0 +1,276 @@
+        Init(label, email, Guid.NewGuid().ToString(), createStringOverTime, logType, logCommonColumns);
+    }
+
+    public LogStore(string label, string email, string sessionID, bool createStringOverTime,
+        LogType logType = LogType.Normal, bool logCommonColumns = true)
+    {
+        Init(label, email, sessionID, createStringOverTime, logType, logCommonColumns);
+    }
```
I guess you are catering to the use case, that someone might want to have a `LogStore` even if they are not using `LoggingManager` by having two separate constructors? or what is the reason?

Vincent-SD replied:
There are 2 constructors in this class to have the possibility of giving a custom sessionID when creating the LogStore object.
If one is given it will be used, otherwise it will be generated. 

I think I'll edit the `GenerateUIDs` [method](https://github.com/Vincent-SD/ArduinoLogger/blob/06cbd9a1d1288205624b9d286d0c12b86799b5c6/Assets/LoggingManager/LoggingManager.cs#L69) to use the `Guid.NewGuid()` function instead of the current manual way of creating the IDs. 
The ID generated with this method is the one used in each log created in the session.

bastianilso replied:
Ah ok. And in which use case do we need to provide a custom sessionID?

Vincent-SD replied:
We need to make sure that every log in the same session has the same ID (meta logs and regular logs for instance) and we have to generate a new one when starting another session. 

In our use case, the ID has to be defined in the LoggingManager and then provided to each LogStore object (using the [constructor](https://github.com/Vincent-SD/ArduinoLogger/blob/06cbd9a1d1288205624b9d286d0c12b86799b5c6/Assets/LoggingManager/LogStore.cs#L41) that has the ID in its parameter).

Actually, the LogStore constructor that generates the sessionID is never used, I just decided to put this here in case one day (again) this class is no longer used with this LoggingManager but in another project.

bastianilso replied:
OK - well, if we by design generate IDs outside LogStore, I think we should just keep a single constructor.
Other projects could always pass `Guid.NewGuid()` directly as an argument as they initialize the LogStore - for the time being I dont think it warrants having two constructors that are otherwise identical.

Vincent-SD replied:
No problem, I'll just delete it then.

bastianilso commented on Assets/LoggingManager/LogStore.cs:
```diff
@@ -0,0 +1,276 @@
+        SessionId = sessionID;
+        this.logType = logType;
+        this.logCommonColumns = logCommonColumns;
+        if (logType == LogType.Normal && logCommonColumns)
+        {
+            logs.Add("Timestamp", new List<string>());
+            logs.Add("Framecount", new List<string>());
+            logs.Add("SessionID", new List<string>());
+            logs.Add("Email", new List<string>());
+        }
+        else if (logType == LogType.Meta)
+        {
+            logs.Add("SessionID", new List<string>());
+            logs.Add("Email", new List<string>());
+        }
```
FYI, I guess right now it is possible to disable `logCommonColumns` but still ask for a `Meta` logType. But I suggest we just remove the `logCommonColumns` altogether. (see my other comment)

bastianilso commented on Assets/LoggingManager/LogStore.cs:
```diff
@@ -0,0 +1,276 @@
+    //Adds a column to the current row of the Logs
+    public void Add(string column, object data)
+    {
+        if (nbLines > 0 && !logs.Keys.Contains(column))
```
Why do we need to check whether `nbLines` is above 0?

Vincent-SD replied:
If the right part of the condition is true, it means that a new header has been added, so we don't check this when logging the first line (when there is 0 lines) because this is the row that defines the headers.

bastianilso replied:
OK, could you add a comment about this above the if statement? Just so it is clear for the future.

bastianilso commented on Assets/LoggingManager/LogStore.cs:
```diff
@@ -0,0 +1,276 @@
+
+    // Converts the values of the parameters (in a "object format") to a string, formatting them to the
+    // correct format in the process.
+    private static string ConvertToString(object arg)
+    {
+        if (arg is float)
+        {
+            return ((float)arg).ToString("0.0000").Replace(",", ".");
+        }
+        if (arg is int)
+        {
+            return arg.ToString();
+        }
+        if (arg is bool)
+        {
+            return ((bool)arg) ? "TRUE" : "FALSE";
+        }
+        if (arg is Vector3)
+        {
+            return ((Vector3)arg).ToString("0.0000").Replace(",", ".");
+        }
+        return arg.ToString();
+    }
```
Do we have a sanitization process e.g. to remove special characters like `;` or `\n` (field separator, etc) inside `LogStore` ? Or is this done by the LoggingManager? Maybe its better to do this in LogStore, perhaps as part of this function. What do you think?

Vincent-SD replied:
I think it's a good idea to do this in LogStore.

bastianilso reviewed (changes requested):
Hi Vincent, Let's start with your first two commits:

Review:
**Adding the old LoggingManager**  - Looks good. Optionally add a commit description "_Arduino Logger will soon support saving data with Logging Manager. The commits adds the LoggingManager version XX."_ Specify in the title which version of LoggingManager you are adding instead of "old". 👴
**Implementation of LogStore class** - Good message and title. Add to the message _"The LogStore class separates log storage from the LoggingManager class itself."_ so it is clear what is the logstore role to the logging manager. 

See my comments and please reply / let me know what you think.

Vincent-SD commented on Assets/LoggingManager/LogStore.cs:
```diff
@@ -0,0 +1,276 @@
+    private LogType logType;
+    private bool logCommonColumns;
+
+    private int nbLines;
```
nb stands for "Number" (in french we usually use this abbreviation 🥖) , so this variable means number of lines. I think i'll rename it "linesCount", it will be more understandable.

bastianilso replied:
@Vincent-SD  alternatively, we should call it "rowCount" - in line with EndRow, currentRowLogs etc.

Vincent-SD commented on Assets/LoggingManager/LogStore.cs:
```diff
@@ -0,0 +1,276 @@
+public class LogStore : MonoBehaviour
+{
+
+    private SortedDictionary<string, List<string>> logs;
```
This is a dictionary that sorts the keys in alphabetic order. 
At first, i wanted to find a dictionary that keeps the order of insertion of the keys but i didn't find one.

The reason is use this datatype is because when we log the string on the fly, we have to make sure that each log is logged in the right order before appending them to the current big string.

Let's imagine that the logs are the following : 
`Millis;framecount;Email\n`

Now, we need to add our logs to the string in the exact order of the headers :
`2547;10;test@test.dk\n`

Thus, if the logs are not added in the right order, they will be reorganized with this dictionary because each logs is added to this dictionary before being converted to a string.

Vincent-SD commented on Assets/LoggingManager/LogStore.cs:
```diff
@@ -0,0 +1,276 @@
+    public string Label { get; set; }
+
+    private LogType logType;
+    private bool logCommonColumns;
```
I did this just to make the LogStore class usable in another project if needed, or if someone wants to log a very specific log which doesn't require these columns. 
It is not a problem to log those columns by default, but I thought it was cleaner to have the possibility of not logging those columns.

If you think that every log **should** contains the common columns, let's remove this feature.

bastianilso replied:
@Vincent-SD It's also about reducing code complexity (avoiding to increase burden of maintaining code logic). We did not give an option to disable this before, and I dont want to introduce this right now.

If you want, you can keep this boolean internally in logStore and set to `true` by default. But I think for the public API of logstore, we should not expose this, for the time being.

bastianilso replied:
@Vincent-SD Actually, let's remove the bool completely. The thing is, with data analysis, we can always filter out data in post-processing - but we cannot magically create missing data. So I'd rather enforce these columns on people - and stand behind our design decisions. (by e.g. documenting in the README that these columns always are logged).

If people really want to avoid logging them, they can just create hacks - none of our own use cases needs this.

Vincent-SD commented on Assets/LoggingManager/LogStore.cs:
```diff
@@ -0,0 +1,276 @@
+using UnityEngine;
+
+
+public enum LogType
+{
+    Normal,
+    Meta
+}
```
In the LoggingManager, the user can't decide to use Meta logs or regular logs. He can only create a meta log by calling the `CreateLog()` [method](https://github.com/Vincent-SD/ArduinoLogger/blob/06cbd9a1d1288205624b9d286d0c12b86799b5c6/Assets/LoggingManager/LoggingManager.cs#L114). 

I think I misunderstood something with the use of the LogType. 
Can the user have access to this ? If so, what is the purpose ? 

Maybe is it possible to keep this enum like this and change the way we write to the file. Instead of having only one method to write all the lines in the file, we could also add a method that adds a custom line to a file.

bastianilso replied:
@Vincent-SD users of LoggingManager don't access LogStore directly. But LogStore has a public API and needs to be maintained by other developers in the future. Imagine a future developer glancing at `LogType` and asking.. "Normal"...what does normal mean actually? What is "Normal"?

My main feedback to you here is, that it is better to use the enum labels, to be descriptive of the _behavior_ of what the options do. What does "Normal LogType" do in practice? What is a normal logtype? What does a "Meta logtype" do in practice? How does it behave?

In addition to a more descriptive label, I think we should add an in-line code comment here describing the behavior in detail. That's all. 🍸

Vincent-SD commented:
I updated the  Pull Request. 
Regarding the enum, i put comments but I didn't change the labels : 
``` c#
public enum LogType
{
    Normal, //Regular logs - logs automatically commons columns like TimeStamp or Framecount
    Meta //Meta logs - logs only one line containing at least sessionID and Email
}
```
I did this because the only difference between the 2 LogType is the common columns that are logged each row (with the Meta type, it is still possible to log more than one row, should I make this impossible ?). 
In this case, the types correspond well to their real behavior : one type for the Regular logs, one type for the Meta logs.

bastianilso commented:
@Vincent-SD Thanks, I will have a look at your update and resolve the conversations above as I see fit.

> with the Meta type, it is still possible to log more than one row, should I make this impossible?

 1. yes, the Meta type should only be able to log one row. That is the whole definition of Meta - that if there is ever written data to a logtype with this type, the existing data is overwritten. Because, the type of data inside a Meta logtype, describes the logging itself - columns cannot take more than 1 value and if they do, things will break in data analysis.

> I did this because the only difference between the 2 LogType is the common columns that are logged each row
> In this case, the types correspond well to their real behavior : one type for the Regular logs, one type for the Meta logs.

"Logged each row" - I think this is much more descriptive of what this enum actually does than "Regular" or "Normal".

So why not call the `Normal` option e.g. `LogType.RowForRow`  or `LogType.LogEachRow` ?
Something` that describes what choosing the type does. I think it tells developers more about the behavior than `LogType.Normal` or `LogType.Regular`.

in a similar way, `Meta` could be called `LogType.OneRowOverwrite` or something like that.

bastianilso reviewed (changes requested):
Review of commit **Implementing the WriteToCsv class**:
 * Commit title is good.
 * Commit message is good - you are missing a T in "he process is currently"...
 * See my comments/questions on the code itself.
 
 :tropical_drink:

Vincent-SD commented on Assets/LoggingManager/WriteToCSV.cs:
```diff
@@ -0,0 +1,90 @@
+using System.Threading;
+using System.Threading.Tasks;
+using UnityEngine;
+using Debug = UnityEngine.Debug;
```
It is because the "Debug" class is present in both *System.Diagnostics* (used for the StopWatches) and *UnityEngine* packages.
In order to use Debug.Log(""), we have to specify from which package to retrieve the function. 
In our case, we are using the Debug function from UnityEngine.

Vincent-SD commented on Assets/LoggingManager/WriteToCSV.cs:
```diff
@@ -0,0 +1,90 @@
+    private string savePath;
+    private string filePath;
+
+    public LogStore LogStore;
```
Yes, just forgot to make it private.

Vincent-SD commented on Assets/LoggingManager/WriteToCSV.cs:
```diff
@@ -0,0 +1,90 @@
+        Init(logStore, savePath);
+    }
+
+    public WriteToCSV(LogStore logStore, string savePath, string fileName)
+    {
+        this.fileName = fileName;
+        Init(logStore, savePath);
+    }
+
+    public WriteToCSV(LogStore logStore, string filePath)
+    {
+        this.filePath = filePath;
+        this.LogStore = logStore;
+    }
+
+    private void Init(LogStore logStore, string path)
+    {
+        this.savePath = path == "" ? Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments) : path;
+        this.filePath = Path.Combine(savePath, fileName);
+        this.LogStore = logStore;
+    }
```
The idea is to give the possibility to create the WriteToCSV object using :

- A full path
- A directory path and a file name
- A directory path, a file prefix and a file extension (in this case the file is named automatically with the date and the log label)

Each way of creating the object corresponds to a constructor.
Currently, only the third one is used, but I thought it would be a good idea to provide those features for later use.
The private Init() method is just here to avoid duplicate code when initializing the object.

bastianilso reviewed (changes requested):
**Review of Adapting LoggingManager:**

@Vincent-SD before I review this commit, I noticed that the other commit "Adding the old LoggingManager" shows changes, which should be inside this commit, fx Line 28:

```
     private Dictionary<string, LogStore> logsList = new Dictionary<string, LogStore>();
```

I think these changes should be moved to the "Adapting LoggingManager" commit instead. "Adding the old LoggingManager" should only contain the original manager in its latest version.

Vincent-SD commented:
I moved all the changes concerning the old arduinoLogger to the "Adding the old LoggingManager" commit.

I also changed the behavior of the meta logs  (OneRowOverwrite) so that only one line can be logged. These changes are in the "Implementation of LogStore class" commit. 

I fixed some bugs I found too 🔧.

bastianilso commented on Assets/LoggingManager/LoggingManager.cs:
```diff
@@ -47,6 +30,10 @@ public class LoggingManager : MonoBehaviour
     [SerializeField]
     private bool enableCSVSave = true;
 
+    [Header("Log string Settings")]
+    [SerializeField]
+    private bool logStringOverTime = true;
+
```
I think a better header could be "Logging Mode".
Add a underneath `Tooltip` explaining the potential performance gains / trade off between setting this on or off.

bastianilso commented on Assets/LoggingManager/LoggingManager.cs:
```diff
@@ -74,45 +61,27 @@ void Awake()
-        sessionID = Md5Sum(System.DateTime.Now.ToString(SystemInfo.deviceUniqueIdentifier + "yyyy:MM:dd:HH:mm:ss.ffff").Replace(" ", "").Replace("/", "").Replace(":", ""));
-        deviceID = SystemInfo.deviceUniqueIdentifier;
+        this.savePath = path;
     }
```
shouldn't `SetSavePath()` also update the save paths in the `LogStore` (logsList) ?

Vincent-SD replied:
Actually the path is given to WriteToCSV objects in their constructor, so the savePath from Logging Manager will be provided to them during their creation.
Those objects are created at the end of the logging process when the user wants to save the data in files.

```c#
    private void SaveToCSV(string label)
    {
        if (!enableCSVSave) return;
        if (logsList.TryGetValue(label, out LogStore logStore))
        {
            WriteToCSV writeToCsv = new WriteToCSV(logStore, savePath, filePrefix, fileExtension);
            writeToCsv.WriteAll();
        }
    }
```
We can see that the savePath is used here (and only here).
Currently, a WriteToCSV object is created for each LogStore object that we want to save in a file.

bastianilso replied:
makes sense, thanks 🌴

bastianilso commented on Assets/LoggingManager/LoggingManager.cs:
```diff
@@ -74,45 +61,27 @@ void Awake()
-        return new Dictionary<string, Dictionary<int, object>>(collections[collectionLabel].log);
+        sessionID = Md5Sum(System.DateTime.Now.ToString(SystemInfo.deviceUniqueIdentifier + "yyyy:MM:dd:HH:mm:ss.ffff").Replace(" ", "").Replace("/", "").Replace(":", ""));
+        deviceID = SystemInfo.deviceUniqueIdentifier;
     }
```
`GenerateUIDs` is no longer needed right? you were using the GUID.newGuid function now, right?

Vincent-SD replied:
Yes, let's remove it.

bastianilso commented on Assets/LoggingManager/LoggingManager.cs:
```diff
@@ -126,220 +95,147 @@ public void SaveLog(string collectionLabel)
     public void CreateLog(string collectionLabel)
     {
-        collections.Add(collectionLabel, new LogCollection());
+        if (collectionLabel == "Meta")
+        {
+            AddMetaCollectionToList();
+            return;
+        }
+        LogStore logStore = new LogStore(collectionLabel, email, sessionID, logStringOverTime);
+        logsList.Add(collectionLabel, logStore);
```
I am not sure we should be checking the `collectionLabel` here. Meta per se, is not supposed to be any kind of magic word. But, if `CreateMetaCollection` is enabled, LoggingManager will by default create a Meta log with DeviceID, SessionID, etc. What is this for exactly?

Vincent-SD replied:
The problem is that the code to log the Meta logs is "hard coded" in the program using the LoggingManager, so I don't really understand the purpose of the ```CreateMetaCollection``` boolean. 
Here is my pov:

>  if CreateMetaCollection is enabled, LoggingManager will by default create a Meta log with DeviceID, SessionID, etc. 

I agree, but if we want to add information to the Meta logs, we have to put these information in the program. For instance in the Arduino Logger, we add the Meta log data in the Arduino.cs class. 
In this case, we would have some lines like this :  
```c#
          LoggingManager.Log("Meta", new Dictionary<string, object>()
          {
              {"PID", pid},
              {"Comment", Comment},
          });
```

In order to add data to Meta logs, we must write the "Meta" string in the ```Log()``` function.
Now that this is written directly in the code, what if we put the ```CreateMetaCollection``` boolean to false ? 

What I wanted to do here is to allow the user to add data to Meta logs very easily while allowing Logging Manager to be compatible with the programs implementing its old version.

Maybe I misunderstood some things about how Meta logs should work in your environment ? 
should the user be able to add meta logs if he wants to, or are the Meta logs only there to log DeviceID, SessionID, etc and nothing else ?

Depending on the case, there should be several other ways of dealing with these meta logs.

bastianilso replied:
@Vincent-SD Thanks for your points. Here is an outline of how meta log behavior is supposed to work:

- When projects import LoggingManager, `CreateMetaCollection` is enabled by default. It means, that LoggingManager by default will create a `*Meta.csv` file, along with any other custom csv file they have logged.
- When `CreateMetaCollection` is enabled, LoggingManager will create a `LogStore` named `Meta` with `LogType.OneRowOverwrite` enabled and add some useful values like deviceID, a timestamp and a sessionID. This should ideally happen on `Start()`.
- Users of LoggingManager are free to write to the `Meta` collection as they please, to add custom values.
- `CreateMetaCollection` can be disabled, because in some cases, importing and merging many csv files in data analysis can be cumbersome. For example, people who use Excel for data analysis may want to just keep things simple and log everything into 1 CSV file. In that case, the `Meta` collection no longer exists and is in principle a name free to be used for other purposes.

To summarize, rather than checking for the name `Meta`, it is better to check for the boolean `CreateMetaCollection`.

bastianilso commented on Assets/LoggingManager/LoggingManager.cs:
```diff
@@ -126,220 +95,147 @@ public void SaveLog(string collectionLabel)
-        foreach (KeyValuePair<string, object> pair in logData)
-        {
-            if (!collections[collectionLabel].log.ContainsKey(pair.Key))
+            if (collectionLabel == "Meta")
```
Again, the collectionLabel should ideally not have some magic words. Can't we just use the same code path as any other log? (given that LogStore holds info about logging behavior already)

bastianilso commented on Assets/LoggingManager/LoggingManager.cs:
```diff
@@ -126,220 +95,147 @@ public void SaveLog(string collectionLabel)
-                collections[collectionLabel].log["Email"] = new Dictionary<int, object>();
-            }
-        }
+    public void TerminateLogRow(string collectionLabel)
+    {
+        logsList[collectionLabel].EndRow();
+    }
```
Why do we need this function in LoggingManager? In the old logging manager, the behavior was the following:

- If the user wants to log a single value, the user calls `Log(string collectionLabel, string columnLabel, object value)`.
- If the user wants to log multiple values, the user calls `Log(string collectionLabel, Dictionary<string, object> logData)`.

And this way LoggingManager could always infer when a line will end - the convention was that calling `Log()` always corresponds to 1 row in a  CSV file - while it is less powerful, i think it makes it simple to understand by design.  It eliminates extra work for the user (data always appear on a new row). In addition it is more accurate because the timestamp with milliseconds will always correspond very closely to the logged event.

Vincent-SD replied:
I didn't know the old LoggingManager was designed like this.
What I did here is a system which can detect when a new line is added.
For instance, let's say we want to log 3 columns : positionX, positionY and positionZ.

* We can log using ```Log(string collectionLabel, string columnLabel, object value)``` 
    * ```c# 
	  loggingManager.Log("Label","positionX",100);
	  loggingManager.Log("Label","positionY",200);
	  loggingManager.Log("Label","positionZ",100);
	   ```
* Or ```Log(string collectionLabel, Dictionary<string, object> logData)``` 
    * ```c# 
	  loggingManager.Log("Label",new Dictionary<string,object>(){
		  {"positionX",100}
		  {"positionY",200}
		  {"positionZ",100}
	  };
	   ```
In our example, our logs would be like this : 
| positionX | positionY | positionZ |
|--|--|--|
| 100 | 200 | 100 |

Now, if we add a new row having the "positionX" as column label, the program will detect that the column has already been logged for this row and will automatically create a new row.

So after this line of code: 
```c# 
loggingManager.Log("Label","positionX",150);
```
The logs will be like this: 
| positionX | positionY | positionZ |
|--|--|--|
| 100 | 200 | 100 |
| 150 |  |  |

The user can also call the ```TerminateLogRow()``` method manually if he wants to manually end a line at a specific point.

There is also the second feature that adds a "NULL" value to each previous rows when adding a new header.

With these two features, we are sure that each row will have the same number of columns without the user having to worry about it (see [this comment](https://github.com/med-material/ArduinoLogger/pull/46#issuecomment-988644241)).

So this behavior is not very complicated to understand by the users and make the code very easy to write : the user only has to call the ```Log()``` method when he wants to log something and that's it, the program manages everything, but if he needs to do something more specific, he can always end a row using the ```TerminateLogRow()``` method. 


After these explanations, do you still think it's better to end a line after each call of ```Log()``` ?
We could also put a boolean in the LoggingManager that, if true, will end the line after each call to ```Log()``` (it would be set to true by default).

bastianilso replied:
@Vincent-SD  The main issue with this logic, is that it makes reading `Timestamp`, `Framecount` etc. for each row problematic  - there is no longer a forced correlation between the reported timestamp and the time which the values were created (it could become arbitrary). It means, that if users are not aware of this, they can run into a lot of trouble in data analysis. 

This is the main reason that the LoggingManager was designed to always terminate the line, after `Log()` is called  - and to force a `Timestamp` on every single row. This behavior was not a limitation/simplification, but a feature.

Unfortunately it is very important to maintain this correlation for any type of temporal data analysis, so I don't think we should put a boolean in `Log()` for this.

Vincent-SD replied:
Okay, thanks for the explanations. I'll do this.

bastianilso commented on Assets/LoggingManager/LoggingManager.cs:
```diff
@@ -126,220 +95,147 @@ public void SaveLog(string collectionLabel)
+            WriteToCSV writeToCsv = new WriteToCSV(logStore, savePath, filePrefix, fileExtension);
+            writeToCsv.WriteAll();
         }
+    }
```
nice.

Maybe throw an error or a `Debug.LogWarning()` if the user is trying to save a logCollection which doesnt exist?

bastianilso commented on Assets/LoggingManager/LoggingManager.cs:
```diff
@@ -126,220 +95,147 @@ public void SaveLog(string collectionLabel)
+        //}
+
+        //connectToMySQL.AddToUploadQueue(logsList[label].log, logsList[label].label);    
+        //connectToMySQL.UploadNow();
```
Does this mean that the current commit completely removes saving to SQL?
I guess you are doing this due to incompatibility issues?

Vincent-SD replied:
Yes, when we created this pull request it wasn't planned to update the MySQL part.
The problem is that the former way of logging data to SQL used a `Dictionary<string, Dictionary<int, object>>`  while now we log everything using strings with the `ExportAll<string>()` [function](https://github.com/Vincent-SD/ArduinoLogger/blob/42af7ff23d76278f7d092e34623cafa8c4abccc3/Assets/LoggingManager/LogStore.cs#L187).

If you want me to update the MySQL part as well, would you prefer that I make the changes in the "Adapting LoggingManager" commit or in a new one ?

bastianilso reviewed (changes requested):
**Review of Adapting LoggingManager:**

- Commit message could use more elaboration - instead of "new system", perhaps you could say `LogStore` and `WriteToCSV` classes.
- See code comments.

hendrikknoche commented:
I think it would make sense to include this comment/justification why the code has the structure as part of the code documentation.
E.g. to avoid future ‘corrections’. Can you include some explanations, Vincent?
Best,
-Hendrik


From: Bastian Ilso ***@***.***>
Reply to: med-material/ArduinoLogger ***@***.***>
Date: Monday, 3 January 2022 at 20:14
To: med-material/ArduinoLogger ***@***.***>
Cc: Subscribed ***@***.***>
Subject: Re: [med-material/ArduinoLogger] Refactoring the LoggingManager - V1.0 (PR #46)


@bastianilso commented on this pull request.

________________________________

In Assets/LoggingManager/LoggingManager.cs<https://github.com/med-material/ArduinoLogger/pull/46#discussion_r777657500>:

> +    public void TerminateLogRow(string collectionLabel)

+    {

+        logsList[collectionLabel].EndRow();

+    }

@Vincent-SD<https://github.com/Vincent-SD> The main issue with this logic, is that it makes reading Timestamp, Framecount etc. for each row problematic - there is no longer a forced correlation between the reported timestamp and the time which the values were created (it could become arbitrary). It means, that if users are not aware of this, they can run into a lot of trouble in data analysis.

This is the main reason that the LoggingManager was designed to always terminate the line, after Log() is called - and to force a Timestamp on every single row. This behavior was not a limitation/simplification, but a feature.

Unfortunately it is very important to maintain this correlation for any type of temporal data analysis, so I don't think we should put a boolean in Log() for this.

—
Reply to this email directly, view it on GitHub<https://github.com/med-material/ArduinoLogger/pull/46#discussion_r777657500>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAPEDH6WSDLKVINUDWYH74TUUHYRJANCNFSM5ITDQTZA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

bastianilso reviewed (approved):
☀
